### ISSUE TITLE
[💡] Documenting API using Swagger / rswag 

### ISSUE THREAD

**Issue by mmd-osm** [openstreetmap/openstreetmap-website#3107]
Documenting APIs on the WIki (https://wiki.openstreetmap.org/wiki/API_v0.6) has a few issues: it isn't maintained consistently, it's not considered authoritative by any means, and not easily consumable.

I looked at [`rswag`](https://github.com/rswag/rswag) as a way to include API documentation right inside this repo, along with a nice UI to test endpoints and a much easier way for API consumers to generate their own code based on the OAS3 spec that comes along with it.

I'm just posting a few screenshots how this could look like. It's mostly auto-generated content that can be enhanced by additional details/descriptions/error status code lists/examples, etc. Maybe this is even good enough to get rid of the Wiki page altogether and direct developers to the swagger docs instead.

#### A list of API 0.6 endpoints, as served via  http://localhost:3000/api-docs/index.html

![sw1](https://user-images.githubusercontent.com/5842757/108423896-bf690d00-7238-11eb-9156-fd72064b5cb8.png)

#### Testing the /api/capabilities endpoints in the browser

![sw2](https://user-images.githubusercontent.com/5842757/108423981-da3b8180-7238-11eb-8873-7044cff993a0.png)

**Comment by simonpoole** [comment]
Being the victim of absolute hopeless swagger generated documentation before (for example maproulette), it is important to note that it is just a system to semi-automatically generate good-looking documentation, not good documentation. 

The later still requires significant effort regardless of the system. Given that 

- the rails port has essentially no code documentation at all
- moving the API documentation from the wiki to the rails port would put the onus on the current maintainers (but see above)
- and significantly raise the bar for contributing to it

I suspect that however well intentioned this is, it basically boils down to not documenting actual current behaviour of the API at all.

The other aspect is that, while a matter of contention among the maintainers and original rails-port devs, you could argue that the API should be specific implementation independent, and as a consequence as a matter of principle the documentation should be independent of any such implementation.

**Comment by tomhughes** [comment]
Yes I'm not convinced about the wisdom of documenting the API in this repository when most of it (at least as used in production) isn't actually implemented here.

**Comment by mmd-osm** [comment]
> moving the API documentation from the wiki to the rails port would put the onus on the current maintainers (but see above)

Valid point, I also thought about it and came to the following conclusion: OSMF board again has [expressed](https://wiki.osmfoundation.org/wiki/Board/Minutes/2021-02-S2S/Discuss_institutionalising_budgeting_and_fund-raising_processes) their willingness to fund topics like "_Andy Allan's "room by room" renovation - API overhauling_" by having a "part-time maintainer to assist Andy Allan.". While people in this repo may find it difficult to exactly grasp what they mean by API renovation, we should definitively leverage the momentum on the API topic.

I believe getting a new paid resource up to speed by improving documentation has a number of benefits:
* Good way to get familiar with the existing Rails code base
* Clear task with no impact on existing code base (-> less effort for maintainers to review potentially breaking pull requests)
* Better documentation would improve overall developer experience

I think @gravitystorm would agree that Rails port will continue to be the center of gravity for functional requirements, database table definitions and a reference implementation, regardless of what we're doing in production.

**Comment by gravitystorm** [comment]
> I think @gravitystorm would agree that Rails port will continue to be the center of gravity for functional requirements, database table definitions and a reference implementation, regardless of what we're doing in production.

Yep, I would agree with that whole-heartedly. I also have plans to make this codebase (slightly) more performant, which will reduce the need for some deployments to add cgimap. So even if OSMF don't use some of the API calls in production, other installations will continue to do so.

> "part-time maintainer to assist Andy Allan."

Thanks for pointing this out; I haven't been involved in any of the discussions but I suspect this has been more focussed on OSMF budget opportunities rather than actually progressing it yet.

> The later still requires significant effort regardless of the system.

@simonpoole I'm not familiar with swagger or the maproulette documentation. Could you give some idea here as to what else would be involved? Is it inline controller method documentation, or more, or something else? Do you have any links to good/bad examples?

I'm quite happy to have the documentation here in the repo, preferably near the code but I'd rather it was under source control than in a free-for-all.

> Testing the /api/capabilities endpoints in the browser

@mmd-osm Is it just get requests that can be tested? I'm concerned because other systems I'm more familiar with that allow online testing like this either have a 'test mode' API, or it's only your own account that you can mess up by playing around while logged in. On the flip side, if all you can test are get requests, it becomes a bit less interesting! Do you have any details on this?

**Comment by mmd-osm** [comment]
> Is it just get requests that can be tested? I'm concerned because other systems I'm more familiar with that allow online testing like this either have a 'test mode' API, or it's only your own account that you can mess up by playing around while logged in. On the flip side, if all you can test are get requests, it becomes a bit less interesting! Do you have any details on this?

You can spin up a _Swagger Petstore_ test frontend at https://editor.swagger.io/  and try out common HTTP verbs (GET, POST, PUT, PATCH, DELETE, ...) + authorization to get some idea how this works. For testing purposes, a few test endpoints could be included, such as http://localhost:3000 and https://master.apis.dev.openstreetmap.org/. Users can then try things out use their own credentials.

**Comment by HarelM** [comment]
+1 for swagger/openAPI generated docs instead of a wiki maintenance, see my comment here: https://github.com/openstreetmap/openstreetmap-website/pull/5943.

In general, now that OSM backend supports Oauth2 token, one can set this token in a swagger UI to send authenticated requests through this UI, which is also great for testing and a great developer experience.

I have an example of my site. which is written in C# and has the swagger UI generated from the code, as can be seen here:
https://israelhiking.osm.org.il/swagger/index.html
Since my site uses OSM as third party authentication you can use your OSM token there to get user data (If you are interested in trying it out).

I have very little experience with ruby, so I wouldn't know where to start in order to integrate it, but in general, you can start by only integrating it and not exposing it until you are happy with what it offers.
In C# you only add annotations to the functions of the controller, and some comments, and swashbuckle (probably the equivalent of rswag) generates the open API spec and the swagger UI straight from the code, no need to maintain anything beside code documentation, which lies next to the code itself, and is really easy to maintain.

But I don't know if this is easy for ruby...
According to chatGPT this is process (which I believe is at least 50% implemented already):

With **Rswag**, you document your API **by writing RSpec tests** that describe each endpoint's behavior, inputs, and outputs.

It's a clever setup because it:

✅ **Documents** your API (OpenAPI-compliant YAML/JSON)  
✅ **Tests** that your API behaves as expected  
✅ **Feeds** that documentation directly into Swagger UI

---

### 📌 Why this approach?

Instead of writing Swagger YAML manually (which can get messy and out of sync), you:

- Write a test for `GET /posts`
- Describe its parameters, responses, and schema
- Run the test, which generates the Swagger JSON
- View the result in Swagger UI

---

### 🧠 Example: Document + test `GET /users`

Let's say you have a `UsersController` with an index action. You can create a test like this:

```ruby
# spec/integration/users_spec.rb
require 'swagger_helper'

RSpec.describe 'Users API', type: :request do
  path '/users' do
    get 'Retrieves all users' do
      tags 'Users'
      produces 'application/json'

      response '200', 'users found' do
        schema type: :array, items: {
          type: :object,
          properties: {
            id: { type: :integer },
            name: { type: :string },
            email: { type: :string }
          },
          required: [ 'id', 'name', 'email' ]
        }

        run_test!
      end
    end
  end
end
```

That does **two things at once**:

- 🎯 Describes the endpoint to Swagger
- ✅ Validates that a 200 response is returned and matches the schema

---

So yeah—you're integrating Swagger by writing **RSpec tests that double as documentation**. It's a two-for-one deal: test coverage + interactive docs.

Want help writing one for a real route in your app? Just tell me which controller/action you'd like to document.

**Comment by tomhughes** [comment]
> you're integrating Swagger by writing **RSpec tests that double as documentation**

So only the one small problem then - we don't use RSpec.

**Comment by mmd-osm** [comment]
The two screenshots I posted initially were based on a similar approach (obviously without using chatgpt back then). The main goal was to declare the API structure, in order to extract it as OpenAPI format (OAS3) later on.

I haven't explored the option to use RSpec for actual tests. I can't comment if there might be some longer term synergies in doing so. In our case it will be mainly adding some declaration on top of what we already have, rather than reusing or rewriting existing tests. This could be done over time, adding further API endpoints as time permits.

**Comment by pablobm** [comment]
Another candidate for this task: https://github.com/openapi-ruby/openapi-ruby. Quick thoughts from the readme, without properly trying it out:
- It supports Minitest, as opposed to only RSpec.
- Schemas declared in a Ruby DSL, separate from the tests themselves.
- Can serve a Swagger UI interface.
- Commits show heavy use of commercial LLM coding tools.
	- Relevant since, as LLM companies start charging users for their actual use, projects that depend on these tools could stop being maintained. Plus other common questions.
- As of these lines: 5 stars, no issues ever filed, about 2 months old.
- Not sure about the facility to test the schema within integration tests. I think I'd simply enable the schema during normal controller/integration API tests and let it crash when the format is not fulfilled. But would need to play with it more.
